File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
def check(x):
    arr
    cur = 0
    for i in range(len(arr)):
        cur += arr[i] // x
    return cur >= n

k, n = map(int, input().split())
arr = []
for _ in range(k):
    arr.append(int(input()))
st = 0
en = 0xFFFFFFFF

while st<en:
    mid = (st+en+1)//2
    if check(mid): st = mid
    else: en = mid-1
print(st)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;
import java.util.*;

public class Main {
	static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder sb = new StringBuilder();
	static StringTokenizer st;
	
	static int[] arr;
	static int n;
	static int k;

	public static void main(String[] args) throws Exception {
		st = new StringTokenizer(input.readLine());
		n = Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		
		arr = new int[n];

		for (int i=0; i<n; i++)
			arr[i] = Integer.parseInt(input.readLine());
		int l = 1;
		int r = Integer.MAX_VALUE;
		
		
		while (l<r) {
			int mid = l + (r-l+1)/2;
			if (check(mid)) {
				l = mid;
			} else{
				r = mid-1;
			}
		}
		
		System.out.println(l);
	}
	
	static boolean check(int mid) {
		int ct = 0;
		for (int a : arr) {
			ct += a/mid;
		}
		return k <= ct;
		
	}
}
